proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,158 @@
+# Proposal: `downloads.getFileHash()`
+
+**Summary**
+
+Allow extensions to query hashes of contents of downloaded files
+programmatically.
+
+**Document Metadata**
+
+**Author:** @bershanskyi
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** @Rob--W
+
+**Created:** 2024-04-08
+
+**Related Issues:** [WECG GitHub issue 401](https://github.com/w3c/webextensions/issues/401)
+
+## Motivation
+
+### Objective
+
+This API will enable extensions to privately and efficiently scan user
+downloads for "Known Unsafe" contents and validate the integrity of "Known Safe"
+files without exposing the raw content of the files.
+
+#### Use Cases
+
+##### Filtering for unwanted downloads
+
+Many antivirus products detect previously seen "safe" and "unsafe" files by
+computing a file content hash and checking it against a set of known hashes.
+These checks require minimal effort on the client side, yet are powerful
+enough to limit the spread of known malware downloads since effectively every
+client gets "inoculated" against a particular download as soon as a single
+antivirus detects a malicious file.
+User-Agents may store file hashes even after the original file is deleted to
+facilitate tracing of malware exposure even if the file is deemed to be
+dangerous after the user initially downloads it.
+
+##### Download integrity assurance
+
+Files can get damaged while they are shared on a file hosting server or NAS.
+While some file formats explicitly support integrity checks like embedded
+checksums or even cryptographic signatures, other files do not. This API
+would allow file managers or extensions to programmatically check the
+integrity of downloaded files if the file source elects to provide them.
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+Schema in Chromium-style IDL:
+```idl
+namespace downloads {
+    callback GetFileHashCallback = void(DOMString hash);
+
+    [supportsPromises, permissions="downloads.getFileHash"]
+    static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+Schema in TypeScript:
+```ts
+namespace downloads {
+  enum FileHashAlgorithm {
+    SHA256 = "SHA-256"
+  }
+
+  getFileHash: (
+    downloadId: number,
+    algorithm?: keyof downloads.FileHashAlgorithm)
+      => Promise<string>;
+}
+```
+
+### New Permissions
+
+No new permission. `downloads.getFileHash()` is honored only if extension has
+`"downloads"` permission and the host access for `DownloadItem.finalUrl`.
```

Comment:
```suggestion
`"downloads"` permission and the host access for `DownloadItem.finalUrl`.
For data:-URLs, host permissions are not required, because the content of the file is already fully visible through the URL.
```

Added extra remark about data:-urls.

Reply:
[Done](https://github.com/w3c/webextensions/compare/591f58d1b321e9bc9682696d64e2df4bcb75c4c1..ddf0ca8fb4737dcd78a9b58409ea6cf7d7c7c901)